### PR TITLE
[13.0][ADD] l10n_th_withholding_tax_cert_multi

### PR DESCRIPTION
**Pull request by ps-tubtim**
This module allows user to create Multi Withholding Tax Certificate from any payment or journal entry document that contains accounting entry on withholding tax.

Note: 1 payment/journal entry : 1 withholding tax cert.

### CODE REVIEW

File: l10n_th_withholding_tax_cert_multi/README.rst [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+.. contents::
+   :local:
+
+Usage
+=====
+
+From supplier payment that has already withheld tax,
+
+* Click action Create Multi Withholding Tax Cert., this bring user to new Withholding Tax Certificates
```
Comment (kittiu) [inline l10n_th_withholding_tax_cert_multi/README.rst]: `Create Multi Withholding Tax Cert` -> `Create Withholding Tax Certs`

File: l10n_th_withholding_tax_cert_multi/README.rst [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+* Fill in required information for Withholding Tax Certificate
```
Comment (kittiu) [inline l10n_th_withholding_tax_cert_multi/README.rst]:
```suggestion
* Fill in required information on each Withholding Tax Certificate
```

File: l10n_th_withholding_tax_cert_multi/README.rst [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+* User can also print out Withholding Tax Cert. as PDF
```
Comment (kittiu) [inline l10n_th_withholding_tax_cert_multi/README.rst]:
```suggestion
* User can also print out multiple Withholding Tax Certs in a single PDF
```

File: l10n_th_withholding_tax_cert_multi/README.rst [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    :target: https://translation.odoo-community.org/projects/l10n-thailand-13-0/l10n-thailand-13-0-l10n_th_withholding_tax_cert_multi
+    :alt: Translate me on Weblate
+.. |badge5| image:: https://img.shields.io/badge/runbot-Try%20me-875A7B.png
+    :target: https://runbot.odoo-community.org/runbot/238/13.0
+    :alt: Try me on Runbot
+
+|badge1| |badge2| |badge3| |badge4| |badge5| 
+
+This module allow user to create Multi Withholding Tax Certificate from any payment
+or journal entry document that contain accounting entry on withholding tax.
```
Comment (kittiu) [inline l10n_th_withholding_tax_cert_multi/README.rst]:
```suggestion
This module extend capability to create multiple Withholding Tax Certificates at the same time,
from any eligible payments or journal entries.
```

File: l10n_th_withholding_tax_cert_multi/__manifest__.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+    "name": "Thai Localization - Multi Withholding Tax Certificate",
+    "version": "13.0.1.0.0",
+    "author": "Ecosoft, Odoo Community Association (OCA)",
+    "license": "AGPL-3",
+    "website": "https://github.com/OCA/l10n-thailand",
+    "category": "Localization / Accounting",
+    "depends": ["l10n_th_withholding_tax_cert"],
+    "data": ["wizard/create_withholding_tax_cert.xml"],
+    "installable": True,
+    "development_status": "alpha",
```
Comment (kittiu) [inline l10n_th_withholding_tax_cert_multi/__manifest__.py]:
```suggestion
    "development_status": "Alpha",
```
Now, ocabot need capital letter.

File: l10n_th_withholding_tax_cert_multi/wizard/create_withholding_tax_cert.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
+# Copyright 2020 Ecosoft Co., Ltd (https://ecosoft.co.th/)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html)
+
+from odoo import _, fields, models
+from odoo.exceptions import UserError
+
+INCOME_TAX_FORM = [
```
Comment (kittiu) [inline l10n_th_withholding_tax_cert_multi/wizard/create_withholding_tax_cert.py]: Please use this by import form here,
https://github.com/OCA/l10n-thailand/blob/13.0/l10n_th_withholding_tax_cert/models/withholding_tax_cert.py#L8

I.e.,
```
from odoo.addons.l10n_th_withholding_tax_cert.models.withholding_tax_cert import INCOME_TAX_FORM
```

File: l10n_th_withholding_tax_cert_multi/wizard/create_withholding_tax_cert.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    ("pnd1", "PND1"),
+    ("pnd3", "PND3"),
+    ("pnd3a", "PND3a"),
+    ("pnd53", "PND53"),
+]
+
+WHT_CERT_INCOME_TYPE = [
```
Comment (kittiu) [inline l10n_th_withholding_tax_cert_multi/wizard/create_withholding_tax_cert.py]: This too, should import from existing module.

**Comment by ps-tubtim** [comment]
@kittiu Thank you for your review. All comments are updated.

File: l10n_th_withholding_tax_cert_multi/README.rst (hunk 1 above)
Comment (ps-tubtim) [reply l10n_th_withholding_tax_cert_multi/README.rst]: Edited.

File: l10n_th_withholding_tax_cert_multi/README.rst (hunk 2 above)
Comment (ps-tubtim) [reply l10n_th_withholding_tax_cert_multi/README.rst]: Edited.

File: l10n_th_withholding_tax_cert_multi/README.rst (hunk 3 above)
Comment (ps-tubtim) [reply l10n_th_withholding_tax_cert_multi/README.rst]: Edited.

File: l10n_th_withholding_tax_cert_multi/README.rst (hunk 4 above)
Comment (ps-tubtim) [reply l10n_th_withholding_tax_cert_multi/README.rst]: Edited.

File: l10n_th_withholding_tax_cert_multi/wizard/create_withholding_tax_cert.py (hunk 6 above)
Comment (ps-tubtim) [reply l10n_th_withholding_tax_cert_multi/wizard/create_withholding_tax_cert.py]: Edited.

File: l10n_th_withholding_tax_cert_multi/wizard/create_withholding_tax_cert.py (hunk 7 above)
Comment (ps-tubtim) [reply l10n_th_withholding_tax_cert_multi/wizard/create_withholding_tax_cert.py]: Edited.

File: l10n_th_withholding_tax_cert_multi/__manifest__.py (hunk 5 above)
Comment (ps-tubtim) [reply l10n_th_withholding_tax_cert_multi/__manifest__.py]: Edited.

File: l10n_th_withholding_tax_cert_multi/wizard/create_withholding_tax_cert.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        ctx.update(
+            {
+                "wt_account_ids": self.wt_account_ids.ids,
+                "wt_cert_income_type": self.wt_cert_income_type,
+            }
+        )
+        wht_certs = []
+        for active_id in active_ids:
+            doc = self.env[active_model].browse(active_id)
+            vals = {
```
Comment (kittiu) [inline l10n_th_withholding_tax_cert_multi/wizard/create_withholding_tax_cert.py:39]: Create WHT cert from Payments -> OK
Create WHT cert from Move -> Error no supplier_partner_id

File: l10n_th_withholding_tax_cert_multi/wizard/create_withholding_tax_cert.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+                "name": doc.name,
+                "supplier_partner_id": doc.partner_id.id,
```
Comment (kittiu) [inline l10n_th_withholding_tax_cert_multi/wizard/create_withholding_tax_cert.py]:
```suggestion
```

File: l10n_th_withholding_tax_cert_multi/wizard/create_withholding_tax_cert.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+                "income_tax_form": self.income_tax_form,
+            }
+            if active_model == "account.move":
+                vals.update({"move_id": active_id, "date": doc.date})
```
Comment (kittiu) [inline l10n_th_withholding_tax_cert_multi/wizard/create_withholding_tax_cert.py]:
```suggestion
                partner = object.line_ids.filtered(
                    lambda l: l.account_id.id in self.wt_account_ids.ids
                ).mapped("partner_id")
                if not len(partner) == 1:
                    raise UserError(_("No single partner to create WHT Cert"))
                vals.update({"move_id": active_id,
                                     "date": doc.date,
                                     "supplier_partner_id": partner.id})
```

File: l10n_th_withholding_tax_cert_multi/wizard/create_withholding_tax_cert.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+                    lambda l: l.account_id.id in self.wt_account_ids.ids
+                )
+                if not payment_wt:
+                    raise UserError(
+                        _(
+                            "Can not create withholding tax cert. Selected account "
+                            "does not match with Journal Items."
+                        )
+                    )
+                vals.update({"payment_id": active_id, "date": doc.payment_date})
```
Comment (kittiu) [inline l10n_th_withholding_tax_cert_multi/wizard/create_withholding_tax_cert.py]:
```suggestion
                vals.update({"payment_id": active_id, "date": doc.payment_date, "supplier_partner_id": object.partner_id.id})
```

**Comment by kittiu** [comment]
Also, the new Cert. will not have the % yet. I understand if create from JE, it won't have. But if create from Payment, it could, right?

Please check existing WHT cert, how it works and apply it auto as we much as we can.

**Comment by ps-tubtim** [comment]
> Also, the new Cert. will not have the % yet. I understand if create from JE, it won't have. But if create from Payment, it could, right?
> 
> Please check existing WHT cert, how it works and apply it auto as we much as we can.

@kittiu I try to create the new Cert. from Payment But the new Cert. not have the %. Could you show me your step?

WHT

**Comment by ps-tubtim** [comment]
@kittiu I already edited following all your review.

File: l10n_th_withholding_tax_cert_multi/wizard/create_withholding_tax_cert.py (hunk 8 above)
Comment (ps-tubtim) [reply l10n_th_withholding_tax_cert_multi/wizard/create_withholding_tax_cert.py:39]: Edited.

File: l10n_th_withholding_tax_cert_multi/wizard/create_withholding_tax_cert.py (hunk 9 above)
Comment (ps-tubtim) [reply l10n_th_withholding_tax_cert_multi/wizard/create_withholding_tax_cert.py]: Edited.

File: l10n_th_withholding_tax_cert_multi/wizard/create_withholding_tax_cert.py (hunk 10 above)
Comment (ps-tubtim) [reply l10n_th_withholding_tax_cert_multi/wizard/create_withholding_tax_cert.py]: Edited.

File: l10n_th_withholding_tax_cert_multi/wizard/create_withholding_tax_cert.py (hunk 11 above)
Comment (ps-tubtim) [reply l10n_th_withholding_tax_cert_multi/wizard/create_withholding_tax_cert.py]: Edited.

**Comment by kittiu** [comment]
> > Also, the new Cert. will not have the % yet. I understand if create from JE, it won't have. But if create from Payment, it could, right?
> > Please check existing WHT cert, how it works and apply it auto as we much as we can.

If everything start from Vendor Bill. In Vendor Bill, user can select "WT" in inovice line. Then you make paymnent, WT will also passed on to WHT Cert.

And so, I think if you test from invoices and select WT correctly,  it should be possible if you imitate how single Cert works.

File: l10n_th_withholding_tax_cert_multi/wizard/create_withholding_tax_cert.py [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+        active_model = ctx.get("active_model", False)
+        ctx.update(
+            {
+                "wt_account_ids": self.wt_account_ids.ids,
+                "wt_cert_income_type": self.wt_cert_income_type,
+            }
+        )
+        wht_certs = []
+        for active_id in active_ids:
+            doc = self.env[active_model].browse(active_id)
```
Comment (kittiu) [inline l10n_th_withholding_tax_cert_multi/wizard/create_withholding_tax_cert.py]:
```suggestion
        docs = self.env[active_model].browse(active_ids)
            for doc in docs:
```

Comment (ps-tubtim) [reply l10n_th_withholding_tax_cert_multi/wizard/create_withholding_tax_cert.py]: Edited.

**Comment by kittiu** [comment]
@ps-tubtim please rebase
https://github.com/OCA/l10n-thailand/pull/197

**Comment by kittiu** [comment]
@ps-tubtim please add test too.

**Comment by kittiu** [comment]
Moved to https://github.com/OCA/l10n-thailand/pull/200

**Pull request closed without merging**